File: CfxGenerator/CefExportFunction.vb
'' Copyright (c) 2014-2015 Wolfgang Borgsmüller
'' All rights reserved.
'' 
'' Redistribution and use in source and binary forms, with or without 
'' modification, are permitted provided that the following conditions 
'' are met:
'' 
'' 1. Redistributions of source code must retain the above copyright 
''    notice, this list of conditions and the following disclaimer.
'' 
'' 2. Redistributions in binary form must reproduce the above copyright 
''    notice, this list of conditions and the following disclaimer in the 
''    documentation and/or other materials provided with the distribution.
'' 
'' 3. Neither the name of the copyright holder nor the names of its 
''    contributors may be used to endorse or promote products derived 
''    from this software without specific prior written permission.
'' 
'' THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
'' "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
'' LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS 
'' FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE 
'' COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
'' INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
'' BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS 
'' OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND 
'' ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR 
'' TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE 
'' USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Public Class CefExportFunction
    Implements ISignatureOwner

    Public Class Comparer
        Implements IComparer(Of CefExportFunction)
        Public Function Compare(x As CefExportFunction, y As CefExportFunction) As Integer Implements System.Collections.Generic.IComparer(Of CefExportFunction).Compare
            Return String.Compare(x.Name, y.Name)
        End Function
    End Class

    Public ReadOnly Name As String
    Public ReadOnly Signature As Signature
    Public ReadOnly PrivateWrapper As Boolean
    Public ReadOnly Comments As CommentData

    Public ReadOnly Parent As CefType

    Public ApiIndex As Integer

    Public ReadOnly Platform As CefPlatform

    Public Sub New(parent As CefType, fd As Parser.FunctionData, api As ApiTypeBuilder, platform As CefPlatform)
        Me.Name = fd.Name
        Me.Comments = fd.Comments
        Me.Signature = Signature.Create(Me, fd.Signature, api)
        Me.PrivateWrapper = GeneratorConfig.HasPrivateWrapper(Me.Name)
        Me.Parent = parent
        Me.Platform = platform
    End Sub

    Public Sub New(parent As CefType, fd As Parser.FunctionData, api As ApiTypeBuilder)
        Me.New(parent, fd, api, CefPlatform.Independent)
    End Sub

    Public ReadOnly Property CfxName As String
        Get
            Static _name As String = "cfx_" & Name.Substring(4)
            Return _name
        End Get
    End Property

    Public ReadOnly Property ReturnType As ApiType
        Get
            Return Signature.PublicReturnType
        End Get
    End Property

    Public ReadOnly Property PublicName As String
        Get
            Static _name As String = CSharp.ApplyStyle(Name).Substring(If(Parent Is Nothing, 3, Parent.ClassName.Length)) & If(PrivateWrapper, "Private", "")
            Return _name
        End Get
    End Property

    Public ReadOnly Property ProxyFunctionName As String
        Get
            Static _name As String = CSharp.ApplyStyle(CfxName).Substring(If(Parent Is Nothing, 3, 0))
            Return _name
        End Get
    End Property

    Public Sub EmitPInvokeDeclaration(b As CodeBuilder)
        b.AppendComment(Me.ToString())
        CodeSnippets.EmitPInvokeDelegate(b, CfxName, Signature)
    End Sub

    Public Sub EmitNativeFunction(b As CodeBuilder)

        b.AppendComment(Me.ToString())
        If Platform <> CefPlatform.Independent Then
            b.AppendLine("#ifdef CFX_" & Platform.ToString().ToUpperInvariant())
        End If
        b.BeginBlock(Signature.NativeSignature(CfxName))
        Signature.EmitNativeCall(b, Name)
        b.EndBlock()
        If Platform <> CefPlatform.Independent Then
            b.AppendLine("#else")
            b.AppendLine("#define {0} 0", CfxName)
            b.AppendLine("#endif")
        End If
    End Sub

    Public Sub EmitPublicFunction(b As CodeBuilder)

        b.AppendSummaryAndRemarks(Comments)

        Dim modifiers = If(PrivateWrapper, "private", "public")

        If Signature.ManagedArguments.Length = 0 OrElse Not Signature.ManagedArguments(0).IsThisArgument Then
            modifiers &= " static"
        End If

        b.BeginFunction(Signature.PublicSignature(PublicName), modifiers)
        If Platform <> CefPlatform.Independent Then
            b.AppendLine("CfxApi.CheckPlatformOS(CfxPlatformOS.{0});", Platform.ToString())
        End If

        Signature.EmitPublicCall(b)
        b.EndBlock()

    End Sub

    Public Sub EmitRemoteFunction(b As CodeBuilder)

        b.AppendSummaryAndRemarks(Comments, True)

        If Parent Is Nothing Then
            b.BeginFunction(PublicFunctionName, ReturnType.RemoteSymbol, Signature.RemoteSignature)
            Signature.EmitRemoteCall(b)
        Else
            Dim sig = Signature.RemoteSignature
            If String.IsNullOrWhiteSpace(sig) Then
                sig = "CfrRuntime remoteRuntime"
            Else
                sig = "CfrRuntime remoteRuntime, " & sig
            End If
            b.BeginFunction(PublicFunctionName, ReturnType.RemoteSymbol, sig, "public static")
            Signature.EmitRemoteCall(b)
        End If
        b.EndBlock()

    End Sub


    Public Overrides Function ToString() As String
        Return String.Format("CEF_EXPORT {1} {0}({2});", Name, ReturnType, Signature)
    End Function

    Public ReadOnly Property CallType As CfxCallMode Implements ISignatureOwner.CallMode
        Get
            Return CfxCallMode.FunctionCall
        End Get
    End Property

    Public ReadOnly Property CefName As String Implements ISignatureOwner.CefName
        Get
            Return Name
        End Get
    End Property

    Public ReadOnly Property CfxApiFunctionName As String Implements ISignatureOwner.CfxApiFunctionName
        Get
            Return CfxName
        End Get
    End Property

    Public ReadOnly Property PublicFunctionName As String Implements ISignatureOwner.PublicFunctionName
        Get
            Return PublicName
        End Get
    End Property

    Public ReadOnly Property PropertyName As String Implements ISignatureOwner.PropertyName
        Get
            Return Nothing
        End Get
    End Property

    Public ReadOnly Property RemoteCallId As String Implements ISignatureOwner.RemoteCallId
        Get
            Return PublicClassName & PublicName & "RenderProcessCall"
        End Get
    End Property

    Public ReadOnly Property PublicClassName As String Implements ISignatureOwner.PublicClassName
        Get
            If Parent IsNot Nothing Then
                Return Parent.ClassName
            Else
                Return "CfxRuntime"
            End If
        End Get
    End Property

    Public ReadOnly Property Comments1 As CommentData Implements ISignatureOwner.Comments
        Get
            Return Comments
        End Get
    End Property

    Public ReadOnly Property CefConfig As CefConfigData Implements ISignatureOwner.CefConfig
        Get
            Return Nothing
        End Get
    End Property
End Class

File: CfxGenerator/CefExportFunction.cs
// Copyright (c) 2014-2017 Wolfgang Borgsmüller
// All rights reserved.
// 
// This software may be modified and distributed under the terms
// of the BSD license. See the License.txt file for details.

using System.Collections.Generic;

public class CefExportFunction {

    public readonly string Name;
    public readonly Signature Signature;
    public readonly bool PrivateWrapper;

    public CommentNode Comments { get; private set; }

    public readonly CefType Parent;

    public int ApiIndex;

    public readonly CefPlatform Platform;

    public CefExportFunction(CefType parent, Parser.FunctionNode fd, ApiTypeBuilder api, CefPlatform platform) {
        this.Name = fd.Name;
        this.Comments = fd.Comments;
        this.Signature = Signature.Create(SignatureType.LibraryCall, CefName, CefConfig, CallMode, fd.Signature, api);
        this.PrivateWrapper = GeneratorConfig.HasPrivateWrapper(this.Name);
        this.Parent = parent;
        this.Platform = platform;
    }

    public CefExportFunction(CefType parent, Parser.FunctionNode fd, ApiTypeBuilder api)
        : this(parent, fd, api, CefPlatform.Independent) {
    }

    public string CfxName {
        get { return "cfx_" + Name.Substring(4); }
    }

    public ApiType ReturnType {
        get { return Signature.PublicReturnType; }
    }

    public string PublicName {
        get { return CSharp.ApplyStyle(Name).Substring(Parent == null ? 3 : Parent.ClassName.Length) + (PrivateWrapper ? "Private" : ""); }
    }

    public void EmitPInvokeDeclaration(CodeBuilder b) {
        b.AppendComment(this.ToString());
        CodeSnippets.EmitPInvokeDelegate(b, CfxName, Signature);
    }

    public void EmitNativeFunction(CodeBuilder b) {
        b.AppendComment(this.ToString());
        if(Platform != CefPlatform.Independent) {
            b.AppendLine("#ifdef CFX_" + Platform.ToString().ToUpperInvariant());
        }
        b.BeginBlock(Signature.NativeFunctionHeader(CfxName));
        Signature.EmitNativeCall(b, Name);
        b.EndBlock();
        if(Platform != CefPlatform.Independent) {
            b.AppendLine("#else");
            b.AppendLine("#define {0} 0", CfxName);
            b.AppendLine("#endif");
        }
    }

    public void EmitPublicFunction(CodeBuilder b, string apiClassName) {
        b.AppendSummaryAndRemarks(Comments);

        var modifiers = PrivateWrapper ? "private" : "public";

        if(Signature.ManagedParameters.Length == 0 || !Signature.ManagedParameters[0].IsThisArgument) {
            modifiers += " static";
        }

        b.BeginFunction(Signature.PublicFunctionHeader(PublicName), modifiers);
        if(Platform != CefPlatform.Independent) {
            b.AppendLine("CfxApi.CheckPlatformOS(CfxPlatformOS.{0});", Platform.ToString());
        }

        Signature.EmitPublicCall(b, apiClassName, CfxApiFunctionName);
        b.EndBlock();
    }

    public void EmitRemoteFunction(CodeBuilder b) {
        b.AppendSummaryAndRemarks(Comments, true);

        if(Parent == null) {
            b.BeginFunction(PublicFunctionName, ReturnType.RemoteSymbol, Signature.RemoteParameterList, "public static");
            Signature.EmitRemoteCall(b, RemoteCallId, true);
        } else {
            var sig = Signature.RemoteParameterList;
            if(string.IsNullOrWhiteSpace(sig)) {
                sig = "CfrRuntime remoteRuntime";
            } else {
                sig = "CfrRuntime remoteRuntime, " + sig;
            }
            b.BeginFunction(PublicFunctionName, ReturnType.RemoteSymbol, Signature.RemoteParameterList, "public static");
            Signature.EmitRemoteCall(b, RemoteCallId, true);
        }
        b.EndBlock();
    }

    public override string ToString() {
        return string.Format("CEF_EXPORT {1} {0}({2});", Name, ReturnType, Signature);
    }

    public CfxCallMode CallMode {
        get { return CfxCallMode.FunctionCall; }
    }

    public string CefName {
        get { return Name; }
    }

    public string CfxApiFunctionName {
        get { return CfxName; }
    }

    public string PublicFunctionName {
        get { return PublicName; }
    }

    public string PropertyName {
        get { return null; }
    }

    public string RemoteCallId {
        get { return PublicClassName + PublicName + "RemoteCall"; }
    }

    public string PublicClassName {
        get {
            if(Parent != null) {
                return Parent.ClassName;
            } else {
                return "CfxRuntime";
            }
        }
    }

    public CefConfigNode CefConfig {
        get { return new CefConfigNode(); }
    }
}